EOS-19636:  Uploading object with 2KB metadata size fails

vimalraghwani20 opened the pull request:
# Change Summary
## Problem Statement/Description
_[EOS-19636](https://jts.seagate.com/browse/EOS-19636):_
_Uploading object with 2KB metadata size fails_

## Solution Overview
root cause: 
S3 server adds identifier "x-amz-meta-" in to metatadata key in to header. When the metadata length gets calculated it was also considering the above additional length which was causing the upload object fail. 

Fix:
Subtracted the length of  "x-amz-meta-" from the metadata length.
Corrected the error message for the header or metadata exceed.   

## Unit Test Cases
Performed the object upload with metadata as 2048 on dev VM. Object uploaded successfully.
Performed the object upload with metadata as 2049 on dev VM. Object upload failed with proper error message.
 
<details>
  <summary>Output of UT</summary>
[root@ssc-vm-3160 cortx-s3server]# aws s3 cp LICENSE s3://mp-bkt-test1619002304 --endpoint-url http://s3.seagate.com --metadata "26NyodkBkH=7CNkGWnzzoWuWEPBzUcwNsA49LhPtBfsksTfEYrHmE8JaN9PqsR3RXPtFQk5r8moPDQfartJkUsvPbTyviXnFVWjkfKGfjdW3UaJAFD7BiGlvQxN6YJvpfBokHvKnfFWZWu17OHfKS8nQxC2eQliS05fSEwjZgW8I6vsX800ZTiVmw4eFNcwdNooG9bp5eVIQW4BMn5dxzLEAyd8IcNw8cXYhwXdYmRNa6sAgiX49Awm0K8fR3o25wgAuzs9NbkwQAzFcyQ0RXNZvrIpoe31dWbdVz3yLgXxuuz1vzx2Ntseyk9YSXd5m09HHefyHUpfUDF44u96wcnSqJ3xbxZZN5ngNB2Vzk70yJHyohELUAHj3hOngyvQnrKzAAf75dIugoTKFGljZkVQzEDyYYofBXKolWjxj1NsAxHnSO3L1b5rxqBPX7nBdhQpnJW5aY0gjfKzYCkFxSKGVWOZruPC8Xc9Cha9FiUTK1RnD4DDuXTY6VTtVfqiPnOVB31VvFzyUBBU0DUBFNpFQ2AQNLaRSdSx84o0tcuZT0OhTlol1yG5WBQTfiS96OcE53M1BaDdu53MYCiPdcEkfln84WN5KmHr123WvK8fB3LQsW1g2eXlrHf7rmrwEg4vhLVRvumZ1efQJqtfnrtJZuzjSRa9mNnN845spfJVftdpVugiH8eeEw9QGjQHnycWaM00AH5mLCGu2kJvwvNI72zvm9s5aCtJgjdzchJxOPoPlTYClsJNORg2dX27ODs2uMUIuJ40GGhZ3jcxmnqAS9wc2kAEtUiWTxzQFtdFXH22Rgd8ebTpa44AqpZHoCIxSZODPYsrPyKtOwnaPPwltBxSNwPOxBFVPNUpyVPkOLXlO9UdgcJ6JFY63RX6eSeeIFXzJEiJBeJFeaJzlFZ7DIRwYqrKIgxNk0xW6Mdrk1wMTgzwRhk7TFk6fZjS2d9kWCQLr3bnMwPbqNHKh7JOJe49T7tFc8jTVO8Bpb9wU3JbJBlzTdhBRMz9SY2JTQsSQdtBHho9NsKup67CF4VY1Cr9HgxJh4WnNnr99L18swljHHQOEV35L1jJLjhG5Ki7a5fM5mvHYGBvcuIIPeO28TyEVqPJMSTta7ZATdkTCIY5dSUhX6y80qN3iHWvnCO5kxRfUCEVLaJHQ01g8cIPUYoAEwB43K1XWYVY5eaaG86rIK3FQf7xIvQchJYPF1U5HaaqqSVFgk7zAqdywLs2q2SuIkb1Jm1I7YH5O94PBqNlQpIeu1mWZTVkMj8vX08JqXR8jQWHPMpCjOPAjPSFF60RUUwd2j76F0TWfIov9zOzvGISwSOWXUtMydMHk5rvGvzzgZ0xrp9IDPKz2DSvhRgcRb2jgPXht5YCqXwMfD9BElOomObIdGfSFib3YM0uhiIDkekfePMSr5nxO0bRUr7eeBFm3YvoS6f9wpjaiNK00PHtDS3Pql4VC7sOqh7Lczl6HJOEzwetlob25SLAQiDxiLFQHLHJsIex3Kpf51QVSaal0wr2KpS6arMdrlyUuEe8MHt5SBnCTJtikxCCSHT3XWWZvFRTQ4tjhHy3IZNq9ixkeT9kp9sj1yAFkzcaMM0s72tdKHOryiNGbSTwFymwGvEbVnhnJwDnQGsHzzut55sSOK7FBB4Znk6UO2Iyu4SDstgSLWTOGnTcfeQxtVg6jDYEFpG73OWgLt95aLFj2Cp7pBi8BTA6CKuN9i0hIEhmqmFroYiV2RcknDrNWkOr4lEhDtYgNiVajNsRNAzZ9MAeaZpr2q6FJ4kKBKRydS5MLKO2X5hUG5JkzFiCwKEPh8wrQFeqllrTkRUBka2dcQJnbQ0fjrkNlPgnD9h3S6VTDYvQgKICM7aRcOwmqAcUYZeGLmIhCvZq1o0YlgdEUY1LN76EbvOezh22zGZuX2YTJ5rZbRU7FMC2HINujlqPOyeYNkX8g11dBJg9OKWdjSd9UMB2vQby1yIZKmattDd5tOK9W0NTQfuGPQoX5nwTcADauf0w8G1fVtc37vnCsps1nfB612345678901"
upload failed: ./LICENSE to s3://mp-bkt-test1619002304/LICENSE An error occurred (MetadataTooLarge) when calling the PutObject operation: Your metadata headers exceed the maximum allowed metadata size
[root@ssc-vm-3160 cortx-s3server]# aws s3 cp LICENSE s3://mp-bkt-test1619002304 --endpoint-url http://s3.seagate.com --metadata "26NyodkBkH=7CNkGWnzzoWuWEPBzUcwNsA49LhPtBfsksTfEYrHmE8JaN9PqsR3RXPtFQk5r8moPDQfartJkUsvPbTyviXnFVWjkfKGfjdW3UaJAFD7BiGlvQxN6YJvpfBokHvKnfFWZWu17OHfKS8nQxC2eQliS05fSEwjZgW8I6vsX800ZTiVmw4eFNcwdNooG9bp5eVIQW4BMn5dxzLEAyd8IcNw8cXYhwXdYmRNa6sAgiX49Awm0K8fR3o25wgAuzs9NbkwQAzFcyQ0RXNZvrIpoe31dWbdVz3yLgXxuuz1vzx2Ntseyk9YSXd5m09HHefyHUpfUDF44u96wcnSqJ3xbxZZN5ngNB2Vzk70yJHyohELUAHj3hOngyvQnrKzAAf75dIugoTKFGljZkVQzEDyYYofBXKolWjxj1NsAxHnSO3L1b5rxqBPX7nBdhQpnJW5aY0gjfKzYCkFxSKGVWOZruPC8Xc9Cha9FiUTK1RnD4DDuXTY6VTtVfqiPnOVB31VvFzyUBBU0DUBFNpFQ2AQNLaRSdSx84o0tcuZT0OhTlol1yG5WBQTfiS96OcE53M1BaDdu53MYCiPdcEkfln84WN5KmHr123WvK8fB3LQsW1g2eXlrHf7rmrwEg4vhLVRvumZ1efQJqtfnrtJZuzjSRa9mNnN845spfJVftdpVugiH8eeEw9QGjQHnycWaM00AH5mLCGu2kJvwvNI72zvm9s5aCtJgjdzchJxOPoPlTYClsJNORg2dX27ODs2uMUIuJ40GGhZ3jcxmnqAS9wc2kAEtUiWTxzQFtdFXH22Rgd8ebTpa44AqpZHoCIxSZODPYsrPyKtOwnaPPwltBxSNwPOxBFVPNUpyVPkOLXlO9UdgcJ6JFY63RX6eSeeIFXzJEiJBeJFeaJzlFZ7DIRwYqrKIgxNk0xW6Mdrk1wMTgzwRhk7TFk6fZjS2d9kWCQLr3bnMwPbqNHKh7JOJe49T7tFc8jTVO8Bpb9wU3JbJBlzTdhBRMz9SY2JTQsSQdtBHho9NsKup67CF4VY1Cr9HgxJh4WnNnr99L18swljHHQOEV35L1jJLjhG5Ki7a5fM5mvHYGBvcuIIPeO28TyEVqPJMSTta7ZATdkTCIY5dSUhX6y80qN3iHWvnCO5kxRfUCEVLaJHQ01g8cIPUYoAEwB43K1XWYVY5eaaG86rIK3FQf7xIvQchJYPF1U5HaaqqSVFgk7zAqdywLs2q2SuIkb1Jm1I7YH5O94PBqNlQpIeu1mWZTVkMj8vX08JqXR8jQWHPMpCjOPAjPSFF60RUUwd2j76F0TWfIov9zOzvGISwSOWXUtMydMHk5rvGvzzgZ0xrp9IDPKz2DSvhRgcRb2jgPXht5YCqXwMfD9BElOomObIdGfSFib3YM0uhiIDkekfePMSr5nxO0bRUr7eeBFm3YvoS6f9wpjaiNK00PHtDS3Pql4VC7sOqh7Lczl6HJOEzwetlob25SLAQiDxiLFQHLHJsIex3Kpf51QVSaal0wr2KpS6arMdrlyUuEe8MHt5SBnCTJtikxCCSHT3XWWZvFRTQ4tjhHy3IZNq9ixkeT9kp9sj1yAFkzcaMM0s72tdKHOryiNGbSTwFymwGvEbVnhnJwDnQGsHzzut55sSOK7FBB4Znk6UO2Iyu4SDstgSLWTOGnTcfeQxtVg6jDYEFpG73OWgLt95aLFj2Cp7pBi8BTA6CKuN9i0hIEhmqmFroYiV2RcknDrNWkOr4lEhDtYgNiVajNsRNAzZ9MAeaZpr2q6FJ4kKBKRydS5MLKO2X5hUG5JkzFiCwKEPh8wrQFeqllrTkRUBka2dcQJnbQ0fjrkNlPgnD9h3S6VTDYvQgKICM7aRcOwmqAcUYZeGLmIhCvZq1o0YlgdEUY1LN76EbvOezh22zGZuX2YTJ5rZbRU7FMC2HINujlqPOyeYNkX8g11dBJg9OKWdjSd9UMB2vQby1yIZKmattDd5tOK9W0NTQfuGPQoX5nwTcADauf0w8G1fVtc37vnCsps1nfB61234567890"
upload: ./LICENSE to s3://mp-bkt-test1619002304/LICENSE
[root@ssc-vm-3160 cortx-s3server]#

</details>

knrajnambiar76 commented on server/s3_put_object_action.cc:
```diff
@@ -152,7 +152,7 @@ void S3PutObjectAction::validate_put_request() {
 
     s3_put_action_state = S3PutObjectActionState::validationFailed;
-    set_s3_error("BadRequest");
+    set_s3_error("MetadataTooLarge");
```
Can you please verify that in case if the header length crosses 8K for PUT request appropriate error ( MetadataTooLarge) is returned, currently its BadRequest
s3_put_chunk_upload_object_action.cc, s3_put_multiobject_action.cc, s3_put_object_action.cc

vimalraghwani20 replied:
Fixed.

vimalraghwani20 replied:
fixed

amitwac2608 reviewed (changes requested):
Please add proper UT and ST

vimalraghwani20 commented:
@amitwac2608 
I have fixed codecy issue.
Also, as discussed with @knrajnambiar76  we have UT's already in place and I have fixed that too.
As of now, we dont need to add STs for this PR.

vimalraghwani20 commented:
pre-merge job: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/880/console
